src/quicklist.c

```diff
@@ -1168,28 +1195,89 @@ quicklistIter *quicklistGetIterator(const quicklist *quicklist, int direction) {
                                          const int direction,
-                                         const long long idx) {
-    quicklistEntry entry;
+                                         const long long idx)
+{
+    quicklistNode *n;
+    unsigned long long accum = 0;
+    unsigned long long index;
+    int forward = idx < 0 ? 0 : 1; /* < 0 -> reverse, 0+ -> forward */
 
-    if (quicklistIndex(quicklist, idx, &entry)) {
-        quicklistIter *base = quicklistGetIterator(quicklist, direction);
-        base->zi = NULL;
-        base->current = entry.node;
-        base->offset = entry.offset;
-        return base;
+    index = forward ? idx : (-idx) - 1;
+    if (index >= quicklist->count)
+        return NULL;
+
+    /* Seek in the other direction if that way is shorter. */
+    int seek_forward = forward;
+    unsigned long long seek_index = index;
+    if (index > (quicklist->count - 1) / 2) {
+        seek_forward = !forward;
+        seek_index = quicklist->count - 1 - index;
+    }
+
+    n = seek_forward ? quicklist->head : quicklist->tail;
+    while (likely(n)) {
+        if ((accum + n->count) > seek_index) {
+            break;
     } else {
+            D("Skipping over (%p) %u at accum %lld", (void *)n, n->count,
+              accum);
+            accum += n->count;
+            n = seek_forward ? n->next : n->prev;
+        }
+    }
+
+    if (!n)
         return NULL;
+
+    /* Fix accum so it looks like we seeked in the other direction. */
+    if (seek_forward != forward) accum = quicklist->count - n->count - accum;
+
+    D("Found node: %p at accum %llu, idx %llu, sub+ %llu, sub- %llu", (void *)n,
+      accum, index, index - accum, (-index) - 1 + accum);
+
+    quicklistIter *iter = quicklistGetIterator(quicklist, direction);
+    iter->current = n;
+    if (forward) {
+        /* forward = normal head-to-tail offset. */
+        iter->offset = index - accum;
+    } else {
+        /* reverse = need negative offset for tail-to-head, so undo
+         * the result of the original index = (-idx) - 1 above. */
+        iter->offset = (-index) - 1 + accum;
     }
+
+    return iter;
 }
 
 /* Release iterator.
  * If we still have a valid current node, then re-encode current node. */
 void quicklistReleaseIterator(quicklistIter *iter) {
+    if (!iter) return;
     if (iter->current)
         quicklistCompress(iter->quicklist, iter->current);
 
+#ifdef REDIS_TEST
+    unsigned int index = 0;
+    quicklistNode *node = iter->quicklist->head;
+    while(node != NULL) {
+        assert(node->recompress != 1);
+
+        /* Verify that quicklist node outside the compress range
+         * are not compressed. */
+        unsigned compress = iter->quicklist->compress;
+        if (compress == 0 || index < compress ||
+            (iter->quicklist->len - index - 1) < compress)
+        {
+            assert(node->encoding != QUICKLIST_NODE_ENCODING_LZF);
+        }
+
+        index++;
+        node = node->next;
+    }
+#endif
```

Comment:
i don't like this.
so far, the REDIS_TEST define used to specify if additional test code is compiled, but it didn't mean that redis itself has overheads for testing.
please find another way to achieve this.
maybe some boolean flag inside the iterator or a global that's enabled by the test code.

Reply:
Have removed it, stripping the code that matt has implemented to verify it.